api/indexer/tx_indexer.go

```diff
@@ -14,30 +13,58 @@ import (
 	"github.com/ava-labs/hypersdk/chain"
 	"github.com/ava-labs/hypersdk/consts"
 	"github.com/ava-labs/hypersdk/fees"
+	"github.com/ava-labs/hypersdk/vm"
+	"github.com/ava-labs/hypersdk/vm/event"
 )
 
 var (
-	_ AcceptedSubscriber = (*TxDBIndexer)(nil)
-	_ AcceptedSubscriber = (*NoOpTxIndexer)(nil)
-
 	failureByte = byte(0x0)
 	successByte = byte(0x1)
+
+	_ event.SubscriptionFactory[*chain.StatelessBlock] = (*subscriptionFactory)(nil)
+	_ event.Subscription[*chain.StatelessBlock]        = (*txDBIndexer)(nil)
 )
 
-type TxIndexer interface {
-	AcceptedSubscriber
-	GetTransaction(txID ids.ID) (bool, int64, bool, fees.Dimensions, uint64, error)
+func WithIndexer[T vm.Controller](
+	db database.Database,
+	name string,
+	path string,
+) vm.Option[T] {
+	indexer := &txDBIndexer{db: db}
+	subscriptionFactory := &subscriptionFactory{
+		indexer: indexer,
+	}
+	apiFactory := &apiFactory{
+		path:    path,
+		name:    name,
+		indexer: indexer,
+	}
+
+	options := []vm.Option[T]{
+		vm.WithBlockSubscriptions[T](subscriptionFactory),
+		vm.WithVMAPIs[T](apiFactory),
```

Comment:
I feel like this looks weird (i.e you can now re-specify options and they append onto each other) ... but it seems to be better than the alternatives we discussed so I don't want to block on this + it seems good enough for now.